src/main/java/jenkins/plugins/gerrit/api/ssh/GerritApiSSH.java

```diff
@@ -0,0 +1,139 @@
+import com.google.gerrit.extensions.restapi.RestApiException;
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.urswolfer.gerrit.client.rest.http.changes.ChangesParser;
+import com.urswolfer.gerrit.client.rest.http.changes.ReviewResultParser;
+import org.apache.commons.io.IOUtils;
+import org.apache.sshd.client.SshClient;
+import org.apache.sshd.client.channel.ChannelExec;
+import org.apache.sshd.client.future.ConnectFuture;
+import org.apache.sshd.client.session.ClientSession;
+import org.eclipse.jgit.transport.URIish;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+public class GerritApiSSH extends GerritApi.NotImplemented {
+
+    private URIish gerritApiUrl;
+    private final int timeout;
+    private SshClient sshClient;
+    private Gson gson = new Gson();
+    private ChangesParser changesParser = new ChangesParser(gson);
+    private ReviewResultParser reviewResultParser = new ReviewResultParser(gson);
+
+    public GerritApiSSH(SshClient sshClient, URIish gerritApiUrl, int timeout) {
+        this.sshClient = sshClient;
+        this.gerritApiUrl = gerritApiUrl;
+        this.timeout = timeout;
+    }
+
+    @Override
+    public Changes changes() {
+        return new ChangesSSH();
+    }
+
+    private JsonElement request(String request) throws IOException {
+        return request(request, null);
+    }
+
+    private JsonElement request(String request, String stdin) throws IOException {
+        sshClient.start();
+        ConnectFuture connectFuture = sshClient.connect(gerritApiUrl.getUser(), gerritApiUrl.getHost(), gerritApiUrl.getPort());
+        connectFuture.await(timeout, TimeUnit.MILLISECONDS);
+        if (!connectFuture.isConnected()) {
+            throw new IOException(String.format("Failed to connect to %s:%s", gerritApiUrl.getHost(), gerritApiUrl.getPort()));
+        }
+        try (ClientSession session = connectFuture.getSession()) {
+            session.auth().await(timeout, TimeUnit.MILLISECONDS);
+            if (!session.isAuthenticated()) {
+                throw new IOException(String.format("Failed to authenticate to %s:%s", gerritApiUrl.getHost(), gerritApiUrl.getPort()));
+            }
+            if (stdin != null) {
+                try (ChannelExec channel = session.createExecChannel(request)) {
+                    channel.open().await(timeout, TimeUnit.MILLISECONDS);
+                    IOUtils.write(stdin, channel.getInvertedIn(),  StandardCharsets.UTF_8);
+                    channel.getInvertedIn().close();
+                    return gson.fromJson(IOUtils.toString(channel.getInvertedOut(),  StandardCharsets.UTF_8), JsonElement.class);
+                }
+            } else {
+                return gson.fromJson(session.executeRemoteCommand(request), JsonElement.class);
+            }
+        }
+    }
+
+    private class ChangesSSH extends Changes.NotImplemented {
```

Comment:
I had a thought about this, maybe it can work... I would have given this a chance.

Provided the entities in restapi and ssh are the same, I do not know gerrit that much...

What if we use the entire implementation of the restapi while overriding only the RestClient interface? Then we get for free the entire logic, while all we need is to hook the RestClient::*Request, based on URL perform the SSH command.

A clean implementation will also help us to merge this into upstream of the restapi implementation so it support both...

What do you think?

Reply:
That might be difficult as there's not really a 1:1 mapping between the Gerrit REST api, and the Gerrit SSH interface. In fact, the difference is often quite huge.

For example, the comment step is implemented completely differently.
In reviews, the following rest path is used:
/changes/{change-id}/revisions/{revision-id}/review
and for drafts:
/changes/{change-id}/revisions/{revision-id}/drafts

while for both the following SSH command is sent:
gerrit review {change-id},{revision-id} --json
and then the payload is streamed to stdin. In the draft/comment case, the comment is a part of a ReviewInput payload.

So, in order to extend the REST API classes, we'd have to quite heavy modifications to make it work, and I'd actually rather have a NotImplementedException when something isn't implemented correctly, than to have it try something that probably won't work.

What do you think?

Reply:
Hi,
I tried to read the gerrit documentation and I cannot find how I make a comment a draft in ssh.
Can you please give an example of the difference?
Is it possible to perform a "gerrit review xxx" and pipe only draft comment wrapped in ReviewInput? And the publish using another cal to "gerrit review xxx"? If so, it is the same sequence of restapi, as if we get /drafts we can create ReviewInput, set the draft within and then pipe.
Another option we can make is to not use the drafts but stack the comments in the context and then build a full response when invoking GerritReview, but I think the 1st option is doable and easier to process.

Reply:
I had the same issue 😄 

But, yes, it is possible to wrap in a ReviewInput like so:

~~~
{
  "comments": { 
    "file.txt": [
      {"line":"1", "path": "file.txt", "message": "adsfsdaf"}
     ]
  }
}
~~~

I'm not entirely sure how you mean with option 1. Maybe I can finish implementing the comment part and we can discuss it anew after? I'm also implementing better error handling so it's taking some time with experimenting back and forth 😄

Reply:
Oh, another issue is that when using SSH, gerrit returns no response body if successful, which unfortunately doesn't work that well with API interface (no matter which method we use)

It only returns a response if something goes wrong.

Reply:
Correction: It won't return a response for commands that take input on stdin. It will however return JSON as expected for non-stdin commands, such as what would be GET requests in the REST equivalent.

Reply:
can we execute a get following a command to retrieve this "missing" output?

Reply:
I understand your point!

Should we parse the path String then and try to deduce from there? I can give it a go tomorrow. The important bit was the actual SSH communication anyway.

And I'll rebase against your patches.

Reply:
yes, I thought to have a parser of URI for operation, change id, revision id, suboperation and based on operation, sub-operation to call a function with change id, revision id as parameters and provide the input object. we can do this for these activities we need for now and then discuss with upstream what is the best approach.

Reply:
Sure! Should be easy enough.

I'm going to bed now though 😄